packages/edit-site/src/components/add-new-template/utils.js

```diff
@@ -86,10 +86,105 @@ const usePublicTaxonomies = () => {
 	}, [ taxonomies ] );
 };
 
+function usePostTypeNeedsUniqueIdentifier( publicPostTypes ) {
+	const postTypeLabels = useMemo( () =>
+		publicPostTypes?.reduce( ( accumulator, { labels } ) => {
+			const singularName = labels.singular_name.toLowerCase();
+			accumulator[ singularName ] =
+				( accumulator[ singularName ] || 0 ) + 1;
+			return accumulator;
+		}, {} )
+	);
+	return useCallback(
+		( { labels, slug } ) => {
+			const singularName = labels.singular_name.toLowerCase();
+			return postTypeLabels[ singularName ] > 1 && singularName !== slug;
+		},
+		[ postTypeLabels ]
+	);
+}
+
+export function usePostTypeArchiveMenuItems() {
+	const publicPostTypes = usePublicPostTypes();
+	const existingTemplates = useExistingTemplates();
+	const needsUniqueIdentifier =
+		usePostTypeNeedsUniqueIdentifier( publicPostTypes );
+	return useMemo(
+		() =>
+			publicPostTypes
+				?.filter(
+					( { slug } ) =>
+						! existingTemplates.some(
+							( existingTemplate ) =>
+								existingTemplate.slug === 'archive-' + slug
+						)
+				)
+				.map( ( postType ) => {
+					let title;
+					if (
+						usePostTypeArchiveMenuItems.SPECIAL_TITLES[
+							postType.slug
+						]
+					) {
+						title =
+							usePostTypeArchiveMenuItems.SPECIAL_TITLES[
+								postType.slug
+							];
+					} else if ( needsUniqueIdentifier( postType ) ) {
+						title = sprintf(
+							// translators: %1s: Name of the post type e.g: "Post"; %2s: Slug of the post type e.g: "book".
+							__( 'Archive: %1$s (%2$s)' ),
+							postType.labels.singular_name,
+							postType.slug
+						);
+					} else {
+						title = sprintf(
+							// translators: %s: Name of the post type e.g: "Post".
+							__( 'Archive: %s' ),
+							postType.labels.singular_name
+						);
+					}
+					return {
+						slug: 'archive-' + postType.slug,
+						description: usePostTypeArchiveMenuItems
+							.SPECIAL_DESCRIPTIONS[ postType.slug ]
+							? usePostTypeArchiveMenuItems.SPECIAL_DESCRIPTIONS[
+									postType.slug
+							  ]
+							: sprintf(
+									// translators: %s: Name of the post type e.g: "Post".
+									__(
+										'Displays an archive with the latests posts of type: %s.'
+									),
+									postType.labels.singular_name
+							  ),
+						title,
+						// `icon` is the `menu_icon` property of a post type. We
+						// only handle `dashicons` for now, even if the `menu_icon`
+						// also supports urls and svg as values.
+						icon: postType.icon?.startsWith( 'dashicons-' )
+							? postType.icon.slice( 10 )
+							: archive,
+					};
+				} ) || [],
+		[ publicPostTypes, existingTemplates, needsUniqueIdentifier ]
+	);
+}
+usePostTypeArchiveMenuItems.SPECIAL_DESCRIPTIONS = {
```

Comment:
Adding props to the hook here seems a bit weird to me.. I'm not sure if it will affect anything with React internals. Do you know?